RockLib.Configuration.ObjectFactory/README.md

```diff
@@ -0,0 +1,392 @@
+
+```c#
+public class Foo
+{
+    public IEnumerable<Bar> Bars { get; set; }
+}
+
+public class Bar
+{
+    public string Baz { get; set; }
+    public int Qux { get; set; }
+}
+```
+
+instead of a configuration like this:
+
+```json
+{
+    "bars": [
+        {
+            "baz": "abc",
+            "qux": 123
+        }
+    ]
+}
+```
+
+can be rewritten like this:
+
+```json
+{
+    "bars": {
+        "baz": "abc",
+        "qux": 123
+    }
+}
+```
+
+### Default Types
+
+Given a type `Foo` that has an abstract property `Bar`:
+
+```c#
+public class Foo
+{
+    public IBar Bar { get; set; }
+}
+
+public interface IBar
+{
+}
+
+public class DefaultBar : IBar
+{
+    public int Baz { get; set; }
+}
+
+We would like to be able to define the `DefaultBar` type as the default type - if the type is not otherwise specified, create an instance of `DefaultBar`. There are two ways of specifying the default type: by property or by type. When specified by property, we want to say: "when setting the `Foo.Bar` property, if the configuration is *not* type-specified, set the property to an instance of `DefaultBar`." When specified by type we want to say: "whenever creating an instance of `IBar`, if the configuration is *not* type-specified, create an instance of `DefaultBar`.
+
+To programmatically set `DefaultBar` as the default type for the `Foo.Bar`, call the `Create` extension method as follows:
+
+```c#
+DefaultValues defaultValues =
+    new DefaultValues().Add(typeof(Foo), nameof(Foo.Bar), typeof(DefaultBar));
+    
+Foo foo = configuration.Create<Foo>(defaultValues: defaultValues);
+```
+
+To programmatically set `DefaultBar` as the default type for the `IBar`, call the `Create` extension method as follows:
+
+```c#
+DefaultValues defaultValues =
+    new DefaultValues().Add(typeof(IBar), typeof(DefaultBar));
+
+var foo = configuration.Create<Foo>(defaultValues: defaultValues);
+```
+
+Default values can also be specified via attributes, so that the extra parameter when in the `Create` extension method can be omitted. 
+
+```c#
+public class Foo
+{
+    [DefaultType(typeof(DefaultBar))]
+    public IBar Bar { get; set; }
+    
+    public IBaz Baz { get; set; }
+}
+
+public interface IBar
+{
+}
+
+public class DefaultBar : IBar
+{
+    public string Qux { get; set; }
+}
+
+[DefaultType(typeof(DefaultBaz))]
+public interface IBaz
+{
+}
+
+public class Baz : IBaz
+{
+    public bool Corge { get; set; }
+}
+```
+
+### Value Converters
+
+RockLib converts most configuration string values to the target type by using the `TypeConverter` obtained by calling `TypeDescriptor.GetConverter(targetType)`. In addition, there is support for target types `Encoding` and `Type`. If value conversions need to be supported for additional types, value converters can be registered. These value converters a functions that have a single string parameter and return the target type. Similar to default types, they can be registered by property or by type.
```

Comment:
"These value converters a functions" should be "These value converters are functions"